src/lexer.coffee

```diff
@@ -824,6 +825,30 @@ isUnassignable = (name, displayName = name) -> switch
 
 exports.isUnassignable = isUnassignable
 
+# `from` isn’t a CoffeeScript keyword, but it behaves like one in `import` and
+# `export` statements (handled above) and in the declaration line of a `for`
+# loop. Try to detect when `from` is a variable identifier and when it is this
+# “sometimes” keyword.
+fromIsKeywordInForDeclaration = (prev) ->
+  if prev[0] is 'IDENTIFIER'
+    # `for i from from`, `for from from iterable`
+    if prev[1] is 'from'
+      prev[1][0] = 'IDENTIFIER'
+      yes
+    # `for i from iterable`
+    yes
+  # `for from…`
+  else if prev[0] is 'FOR'
+    no
+  # `for {from}…`, `for [from]…`
+  else if prev[1] in ['{', '[']
```

Comment:
Nit: `if prev[1] in '{['` would save an array allocation (not really sure if that's meaningful).